pepper/test/apps/util.h

```diff
@@ -1,40 +1,54 @@
 #include <apps/hashing.h>
 #include <boost/stacktrace.hpp>
+#include <openssl/sha.h>
+#include <pedersen_bridge.cpp>
 
 #ifndef ORDERS
 #define ORDERS 1
 #endif
 
 namespace pepper_overrides
 {
-    /**
-     * Dummy pedersen implementation: XOR lhs and rhs bits
-     */
-    void sha(field254 in[SHA_HASH_SIZE], field254 out[256]) {
-        for (size_t index = 0; index < 256; index++) {
-            out[index] += in[index].is_zero() ^ in[index+256].is_zero();
+    void charArrayToFieldArray(unsigned char *in, field254 *out, size_t length) {
+        for (size_t index = 0; index < length; index++) {
+            for (size_t bit = 0; bit < 8; bit++) {
+                out[(index*8) + bit] = (in[index] >> (7 - bit)) % 2;
+            }
+        }
+    }
+
+    void fieldArrayToCharArray(field254 *in, unsigned char *out, size_t length) {
+        for (size_t index = 0; index < length; index++) {
+            out[index/8] += in[index].is_zero() ? 0 : int(pow(2, 7 - (index % 8)));
         }
     }
 
-    /**
-     * Dummy pedersen implementation: Sum of all inputs
-     */
+    void sha(field254 in[SHA_HASH_SIZE], field254 out[256]) {
+        unsigned char charIn[SHA_HASH_SIZE/8] = {0};
+        unsigned char charOut[32] = {0};
+        SHA256_CTX sha256;
+        SHA256_Init(&sha256);
+        fieldArrayToCharArray(in, charIn, SHA_HASH_SIZE);
+        SHA256_Update(&sha256, charIn, SHA_HASH_SIZE/8);
+        SHA256_Final(charOut, &sha256);
+        charArrayToFieldArray(charOut, out, 32);
+    }
+
     void pedersen(field254 in[PEDERSEN_HASH_SIZE], field254 out[2]) {
-        for (size_t index = 0; index < PEDERSEN_HASH_SIZE; index++) {
-            out[0] += in[index];
+        protoboard<FieldT> pb;
+        char charIn[2*PEDERSEN_HASH_SIZE] = {0};
+        for (size_t i = 0; i < PEDERSEN_HASH_SIZE; i++) {
+            charIn[2*i] = in[i].is_zero() ? '0' : '1';
+            charIn[2*i + 1] = ' ';
         }
+        auto gadget = makeGadget(charIn, pb);
+        out[0] = pb.val(gadget.result_x());
+        out[1] = pb.val(gadget.result_y());
     }
 
     void decomposeBits(field254 number, field254* bits, uint32_t offset) {
-        int index = 0;
-        unsigned long nr = number.as_ulong();
-        while (nr > 0) {
-            if(nr%2==0)
-                bits[offset + 253 - index] = field254::zero();
-            else
-                bits[offset + 253 - index] = field254::one();
-            nr = nr/2;
-            index++;
+        for (size_t index = 0; index < number.as_bigint().num_bits(); index++) {
```

Comment:
minor, minor: number.as_bigint() and number.as_bigint().num_bits() could have been stored outside of the for loop.

Reply:
I'd opt for conciseness/readability here given that performance is not that critical for unit-tests and a decent compiler should do this optimisation under the hood.